data_steward/tools/retract_data.py

```diff
@@ -0,0 +1,132 @@
+        for folder_item in to_process_folder_list:
+            print(folder_item)
+    else:
+        if folder in folder_list:
+            to_process_folder_list = [folder]
+            print("Retracting data from the following folders:")
+            print(folder)
+        else:
+            print('Folder %s does not exist in bucket %s. Exiting' % (folder, bucket))
+            return result
+
+    for folder_prefix in to_process_folder_list:
+        print('Processing gs://%s/%s' % (bucket, folder_prefix))
+        # separate cdm from the unknown (unexpected) files
+        found_files = []
+        folder_items = [item['name'].split('/')[1] for item in bucket_items if item['name'].startswith(folder_prefix)]
+        for item in folder_items:
+            # Only retract from CDM or PII files
+            if val._is_cdm_file(item) or val._is_pii_file(item):
+                found_files.append(item)
+
+        print('Found the following files to retract data from:')
+        for file_name in found_files:
+            print(bucket+'/'+folder_prefix+file_name)
+
+        print("Proceed?")
+        response = get_response()
+        if response == "Y":
+            retract(pid, bucket, found_files, folder_prefix, force)
+        elif response.lower() == "n":
+            print("Ignoring folder %s" % folder_prefix)
+            continue
+
+    return result
+
+
+def retract(pid, bucket, found_files, folder_prefix, force):
+    result = []
+    for file_name in found_files:
+        if force:
+            print("Force retracting rows for person_id %s from path %s/%s%s" % (pid, bucket, folder_prefix, file_name))
+            response = "Y"
+        else:
+            # Make sure user types Y to proceed
+            print("Are you sure you want to retract rows for person_id %s from path %s/%s%s?"
+                  % (pid, bucket, folder_prefix, file_name))
+            response = get_response()
+        if response == "Y":
+            # Output and input file content initialization
+            retracted_file_string = StringIO.StringIO()
+            input_file_string = gcs_utils.get_object(bucket, folder_prefix + file_name)
+            input_contents = input_file_string.split('\n')
+
+            # Check if file has person_id in first or second column
+            if file_name in PID_IN_COL1:
+                for input_line in input_contents:
+                    if input_line != '':
+                        if input_line.split(",")[0] != pid:
+                            retracted_file_string.write(input_line + '\n')
+            elif file_name in PID_IN_COL2:
+                for input_line in input_contents:
+                    if input_line != '':
+                        if input_line.split(",")[1] != pid:
+                            retracted_file_string.write(input_line + '\n')
+            # Write result back to bucket
+            result.append(gcs_utils.upload_object(bucket, folder_prefix + file_name, retracted_file_string))
```

Comment:
Should we bypass saving if there are no rows with the PID in question (the file doesn't need to be updated)?

Reply:
Yeah, it looks like an unnecessary extra step. I'll skip re-writing/uploading for files containing no rows with the PID.